llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp

```diff
@@ -2039,6 +2039,11 @@ bool NVPTXDAGToDAGISel::tryBFE(SDNode *N) {
       Val = AndLHS;
       Start = CurDAG->getTargetConstant(ShiftAmt, DL, MVT::i32);
       Len = CurDAG->getTargetConstant(NumBits, DL, MVT::i32);
+
+      // If we have a arithmetic right shift, we need to use the signed bfe
+      // variant only if the MSB is copied as part of the mask
```

Comment:
This could use some editing. The comment explains what we do, but not why.

E.g. something along the lines of "if pre-shift AND includes the sign bit in the bitfield, we must use signed BFE to replicate that bit during bitfield extraction. If the sign bit is not part of the mask, unsigned BFE will zero out upper bits of the result"